### PR TITLE
[API][SIMS #1371] Populate auditing fields to DB

### PR DESCRIPTION
**Pull request by dheepak-aot**
## Populate auditing fields

- [x] For institution APIs, add and ensure API calls to change data (PATCH, POST, PUT,etc) are providing the audit user as creator/modifier
- [x] For student APIs, add and ensure API calls to change data (PATCH, POST, PUT,etc) are providing the audit user as creator/modifier
- [x] For supporting users APIs, add and ensure API calls to change data (PATCH, POST, PUT,etc) are providing the audit user as creator/modifier
- [x] For aest APIs, add and ensure API calls to change data (PATCH, POST, PUT,etc) are providing the audit user as creator/modifier
- [x] Refactored student notes and restrictions to use transaction and removed the existing deprecated method (saveStudentNote)

### CODE REVIEW

File: sources/packages/api/src/route-controllers/institution-locations/institution-location.institutions.controller.ts [hunk 1]
```diff
@@ -111,8 +112,13 @@ export class InstitutionLocationInstitutionsController extends BaseController {
   async update(
     @Param("locationId") locationId: number,
     @Body() payload: InstitutionLocationPrimaryContactAPIInDTO,
+    @UserToken() userToken: IInstitutionUserToken,
   ): Promise<void> {
-    this.locationService.updateLocationPrimaryContact(payload, locationId);
+    this.locationService.updateLocationPrimaryContact(
```
Comment (ann-aot) [inline sources/packages/api/src/route-controllers/institution-locations/institution-location.institutions.controller.ts:117]: await?

File: sources/packages/api/src/route-controllers/note/note.aest.controller.ts [hunk 2]
```diff
@@ -140,10 +140,11 @@ export class NoteAESTController extends BaseController {
     if (!student) {
       throw new NotFoundException("Student not found.");
     }
-    const studentNote = transformToNoteEntity(payload, userToken.userId);
```
Comment (ann-aot) [inline sources/packages/api/src/route-controllers/note/note.aest.controller.ts:143]: for my info, why `this transformToNoteEntity` is removed

File: sources/packages/api/src/route-controllers/restriction/restriction.aest.controller.ts [hunk 3]
```diff
@@ -187,23 +187,14 @@ export class RestrictionAESTController extends BaseController {
         "The given restriction type is not Provincial. Only provincial restrictions can be added.",
       );
     }
-    // TODO Use transaction to save the restriction and save student note alongside.
-    const updatedRestriction =
+    const addedRestriction =
```
Comment (ann-aot) [inline sources/packages/api/src/route-controllers/restriction/restriction.aest.controller.ts:190]: newRestrcition?

File: sources/packages/api/src/route-controllers/note/note.aest.controller.ts (hunk 2 above)
Comment (dheepak-aot) [reply sources/packages/api/src/route-controllers/note/note.aest.controller.ts:143]: The transformToNoteEntity utility creates note entity model for the sake of saveStudentNote  service method which is deprecated and has been replaced.

File: sources/packages/api/src/route-controllers/restriction/restriction.aest.controller.ts [hunk 4]
```diff
@@ -352,22 +335,22 @@ export class RestrictionAESTController extends BaseController {
...
     // Mapping the note added for restriction to institution notes
     if (updatedRestriction.restrictionNote) {
-      institutionNote = await this.institutionService.saveInstitutionNote(
+      await this.institutionService.saveInstitutionNote(
         institutionId,
         updatedRestriction.restrictionNote,
       );
     }
-    return { id: institutionNote.id };
+    return { id: updatedRestriction.id };
```
Comment (ann-aot) [inline sources/packages/api/src/route-controllers/restriction/restriction.aest.controller.ts:353]: 👍

File: sources/packages/api/src/services/designation-agreement/designation-agreement.service.ts [hunk 5]
```diff
@@ -310,17 +321,26 @@ export class DesignationAgreementService extends RecordDataModelService<Designat
...
+        // When a designation location already exist assign the id
+        // and audit properties.
         if (designationLocation) {
           location.id = designationLocation.id;
-          location.creator = undefined;
-          location.requested = undefined;
+          location.modifier = auditUser;
+          location.updatedAt = designationUpdatedDate;
+        }
+        // When the designation location is not an existing one
```
Comment (ann-aot) [inline sources/packages/api/src/services/designation-agreement/designation-agreement.service.ts:335]: 👍 thanks for the comment

File: sources/packages/api/src/services/education-program-offering/education-program-offering.service.ts [hunk 6]
```diff
@@ -458,13 +458,15 @@ export class EducationProgramOfferingService extends RecordDataModelService<Educ
   ): Promise<void> {
     return this.dataSource.transaction(async (transactionalEntityManager) => {
       // create the note for assessment.
-      const user = {
+      const auditUser = {
```
Comment (ann-aot) [inline sources/packages/api/src/services/education-program-offering/education-program-offering.service.ts:461]: not part of the PR. but starts with caps    ` // Create the note for assessment.`

File: sources/packages/api/src/services/education-program-offering/education-program-offering.service.ts [hunk 7]
```diff
@@ -856,7 +864,7 @@ export class EducationProgramOfferingService extends RecordDataModelService<Educ
         .getRepository(EducationProgramOffering)
         .save([requestedOffering, precedingOffering]);
 
-      // Save applications with new current assessment on approval.
+      // Save applications with new current assessment or status cancelled on approval.
```
Comment (ann-aot) [inline sources/packages/api/src/services/education-program-offering/education-program-offering.service.ts:867]: for my info, could you explain `or status cancelled`?

File: sources/packages/api/src/services/education-program/education-program.service.ts [hunk 8]
```diff
@@ -172,13 +172,17 @@ export class EducationProgramService extends RecordDataModelService<EducationPro
...
     if (!programId) {
       // Institution should never be updated after creation.
       program.institution = { id: institutionId } as Institution;
       program.submittedBy = auditUser;
+      program.submittedDate = now;
       program.creator = auditUser;
+      program.createdAt = now;
     } else {
       program.modifier = auditUser;
+      program.updatedAt = now;
```
Comment (ann-aot) [inline sources/packages/api/src/services/education-program/education-program.service.ts:185]: for else do we need this? it will automatically updated, as you did in draft applications

File: sources/packages/api/src/services/institution/institution.service.ts [hunk 9]
```diff
@@ -768,11 +775,13 @@ export class InstitutionService extends RecordDataModelService<Institution> {
   /**
    * Update institution.
    * @param institutionId
+   * @param auditUserId user who is making the changes.
```
Comment (ann-aot) [inline sources/packages/api/src/services/institution/institution.service.ts:778]: not part of the PR, it would be good if you could add brief details to other comments too

File: sources/packages/api/src/services/student/student.service.ts [hunk 10]
```diff
@@ -590,20 +599,29 @@ export class StudentService extends RecordDataModelService<Student> {
...
+  ): Promise<Note> {
+    return this.dataSource.transaction(async (transactionalEntityManager) => {
+      const note = await this.createStudentNote(
+        studentId,
+        noteType,
+        noteDescription,
+        auditUserId,
+        transactionalEntityManager,
+      );
+      return note;
```
Comment (ann-aot) [inline sources/packages/api/src/services/student/student.service.ts:623]: maybe you can return in line 616

**Review by ann-aot — COMMENTED** [review]
Good Work @dheepak-aot  added some comments. Also Thanks for taking care of old comments 👍

File: sources/packages/api/src/services/education-program-offering/education-program-offering.service.ts (hunk 7 above)
Comment (dheepak-aot) [reply sources/packages/api/src/services/education-program-offering/education-program-offering.service.ts:867]: When an offering change is approved by the ministry, 2 things can possibly take place. If the application status is completed, we create a new student assessment with the new approved offering and start the re-assessment workflow.
If the application is not complete, then the application status is set to cancelled and the student will be asked to re-submit the same application.

I know the comment is slightly confusing to explain what I mentioned above. I have modified the comment like this.

_**Save applications with new current assessment or set status as cancelled on approval.**_

File: sources/packages/api/src/services/education-program/education-program.service.ts (hunk 8 above)
Comment (dheepak-aot) [reply sources/packages/api/src/services/education-program/education-program.service.ts:185]: Right, I missed it.

File: sources/packages/api/src/route-controllers/restriction/restriction.aest.controller.ts [hunk 11]
```diff
@@ -352,22 +335,22 @@ export class RestrictionAESTController extends BaseController {
         "The given restriction type is either not Provincial or not Institution.",
       );
     }
+    // TODO:Modify this method to add restriction and note in a transaction.
```
[the remark this answers is not in the document]
Comment (dheepak-aot) [reply sources/packages/api/src/route-controllers/restriction/restriction.aest.controller.ts:338]: No. Because what I heard from @abschwenker is Institution Restrictions may become obsolete. So I didn't create a ticket for it. Just left a todo.

File: sources/packages/api/src/services/education-program-offering/education-program-offering.service.ts (hunk 7 above)
Comment (ann-aot) [reply sources/packages/api/src/services/education-program-offering/education-program-offering.service.ts]: `Save applications with new current assessment or set application status as cancelled on approval.`

Comment (ann-aot) [reply sources/packages/api/src/services/education-program-offering/education-program-offering.service.ts]: Thanks for the explaination

File: sources/packages/api/src/services/institution/institution.service.ts (hunk 9 above)
Comment (ann-aot) [reply sources/packages/api/src/services/institution/institution.service.ts:778]: Thanks for adding the comments

File: sources/packages/api/src/services/student/student.service.ts [hunk 12]
```diff
@@ -590,20 +600,28 @@ export class StudentService extends RecordDataModelService<Student> {
...
-      where: { id: studentId },
-      relations: { notes: true },
+  async addStudentNote(
+    studentId: number,
+    noteType: NoteType,
+    noteDescription: string,
+    auditUserId: number,
+  ): Promise<Note> {
+    return this.dataSource.transaction(async (transactionalEntityManager) => {
+      return await this.createStudentNote(
```
Comment (ann-aot) [inline sources/packages/api/src/services/student/student.service.ts:617]: i think await is not needed here

File: sources/packages/api/src/services/application/application.service.ts [hunk 13]
```diff
@@ -37,7 +37,6 @@ import { WorkflowActionsService } from "../workflow/workflow-actions.service";
 import { MSFAANumberService } from "../msfaa-number/msfaa-number.service";
 import {
   CustomNamedError,
-  getUTCNow,
```
[the remark this answers is not in the document]
Comment (ann-aot) [reply sources/packages/api/src/services/application/application.service.ts:40]: Thanks for checking it

**Review by ann-aot — APPROVED** [review]
Good Work @dheepak-aot . Thanks for doing the changes 👍 added a minor comment, I think sonar cloud is also complaining the same. and replied a comment.

File: sources/packages/api/src/route-controllers/institution-locations/institution-location.aest.controller.ts [hunk 14]
```diff
@@ -52,7 +58,8 @@ export class InstitutionLocationAESTController extends BaseController {
   async update(
     @Param("locationId") locationId: number,
     @Body() payload: AESTInstitutionLocationAPIInDTO,
+    @UserToken() userToken: IUserToken,
   ): Promise<void> {
-    this.locationService.updateLocation(payload, locationId);
+    this.locationService.updateLocation(payload, locationId, userToken.userId);
```
Comment (guru-aot) [inline sources/packages/api/src/route-controllers/institution-locations/institution-location.aest.controller.ts:63]: I guess we need await here.

**Review by guru-aot — APPROVED** [review]
LGTM, Nice work @dheepak-aot
